feat: add icx-cert tool

roman-kashitsyn opened the pull request:
This change introduces a new command-line tool for debugging certified
HTTP responses.  The tool downloads the document with the specified
URL and pretty-prints the contents of the IC-Certificate header.

hansl commented on icx-cert/src/main.rs:
```diff
@@ -0,0 +1,21 @@
+)]
+enum Command {
+    /// Fetches the specified URL and pretty-prints the certificate.
+    #[clap(name = "pprint")]
```
Non blocking: Just `print` or even no subcommand at all.

roman-kashitsyn replied:
Renamed the command to "print".

hansl commented on icx-cert/src/pprint.rs:
```diff
@@ -0,0 +1,102 @@
+
+/// A fully parsed replica certificate.
+#[derive(Deserialize)]
+struct ReplicaCertificate {
```
Ah yes. I wanted that structure to be maintained by the Agent (and most likely `ic-types` itself), but this is fine I guess.

roman-kashitsyn commented:
Sample output:
```
cargo run -p icx-cert -q -- pprint https://h5aet-waaaa-aaaab-qaamq-cai.raw.ic0.app/index.html
STATUS: 200
ROOT HASH: c430801aca2e94c53eb8d2b9dc74ad28ba004be2408beeabf3022dbc5231ee7d
DATA HASH: 1de2cf06ad5471e7db63e842bf33d97ff9cb5dad0e731de4173d7a2ed319c926
TREE HASH: 3ce000e1a98104c4f6b4655a6d9c508258fdea11b445d0ef88a25215eb35a1fe
SIGNATURE: a9b94d0d6a5ee4eccfc3fefccee91b81e5b41645075b9c437c07c73c156d69122f9881a0883d38b3a7006bfe23ad9f61
CERTIFICATE TREE: HashTree {
    root: Fork(
        Fork(
            Fork(
                Label("canister", Fork(
                    Fork(
                        Fork(
                            Pruned(77fa2e70e208d72fa5c2c40ddd9afa5e669217e5f1b2f78cefce1b39b1064594),
                            Fork(
                                Fork(
                                    Pruned(5a2838de9b66e5d7b8f9364923520d41b98ff974ab54c732355cf047a9ce9408),
                                    Fork(
                                        Fork(
                                            Label("0", Fork(
                                                Fork(
                                                    Label("certified_data", Leaf(0x3ce000e1a98104c4f6b4655a6d9c508258fdea11b445d0ef88a25215eb35a1fe)),
                                                    Pruned(e6fdbef1104a7e4b4620ba1377a5120be7b72eba819c46a7b7b72f7d88a653ff),
                                                ),
                                                Pruned(725ef1ce8efeafae9d37f35d3f997f0f914441a11689809870e49bfae983bd0c),
                                            )),
                                            Pruned(51a81189f3f4f6fe8a2037e98d96ffd1ceab3244cb2e12068765a48f526d75c8),
                                        ),
                                        Pruned(40c0b7d9c992792294cf0626d0b6cd9924ab5eaf14432b995cb5a667974ffaa6),
                                    ),
                                ),
                                Pruned(fa186f05614a9d9b34e9d878e569dd9aebcdc587b83eebe7d2f814c145949c59),
                            ),
                        ),
                        Pruned(c8ee85706ff9c725a1d7e5f5a6cdd10c8b9522472d3e291467e7ee2bc10cfd88),
                    ),
                    Pruned(5bdc70d5498defe69512a3611d43e8ab5b0a69cd01a4e052ae4c2074edc04696),
                )),
                Pruned(aad4156fb9c660611cf6acf11f45941d13bfb7097741b8251ab23a3ccaf52408),
            ),
            Pruned(c0f3ec3037fd3549ebb8fed3f88ed27a36de29bc3827a9ae74ee3dde68785cc8),
        ),
        Fork(
            Pruned(c48b5513eb13e7767efbde6a1e133af12b43608420b3fe2fa85c1160ceb10136),
            Label("time", Leaf(0x80b8bfe0edb5d6c316)),
        ),
    ),
}
TREE:             HashTree {
    root: Label("http_assets", Fork(
        Fork(
            Pruned(1840bdf462a3b0b78541174d296a092fec6baa829ad9fab67ecf5e01160423a2),
            Label("/index.html", Leaf(0x1de2cf06ad5471e7db63e842bf33d97ff9cb5dad0e731de4173d7a2ed319c926)),
        ),
        Pruned(4d2e16d2d76462e17987117a0f3ccae31b0e1994702801202ead74edb02936be),
    )),
}
```

hansl commented on icx-cert/src/pprint.rs:
```diff
@@ -0,0 +1,102 @@
+    let tree: HashTree = parse_base64_cbor(structured_header.tree)?;
+    let cert: ReplicaCertificate = parse_base64_cbor(structured_header.certificate)?;
+
+    println!("STATUS: {}", status);
```
yeah will probably refine this output, but for now this is great!

hansl commented:
lol The DATA and TREE hash are the same as the academy talk this morning :D which makes it great validation.

hansl commented:
Could you show the `time` in ISO 8601 too? Although probably not as useful as it's not the certification time, but the last block time, but I guess it's still nicer than trying to convert it which is a pain (as I learnt yesterday).

nomeata commented:
Can we name it `ic-http-assert-cert` or something more specific? `ic-cert` sounds like it is a tool for all things certification.

hansl commented:
I'd like to have a tool for all thing certification. This can start here and evolve into it.

roman-kashitsyn commented:
> Could you show the `time` in ISO 8601 too? Although probably not as useful as it's not the certification time, but the last block time, but I guess it's still nicer than trying to convert it which is a pain (as I learnt yesterday).

Done.

roman-kashitsyn commented:
@ericswanson-dfinity, could you please merge this PR? I think I addressed all the feedback. There are people who want to use this new shiny tool for debugging.
